Spill join - every LJO unspills

ArturGajowy opened the pull request:
This is what I've got so far. The missing part is synchronizing the LJOs so that they all process the same partition if any ('suspending' the faster LJOs so that they wait till the others finish processing the current partition).

Please have a quick skim so that we have faster feedback on this.

I think the synchronization mentioned above can be done using an atomic integer provided to the LJOs from their factory (number of LJOs processing the current partition) and another one for keeping the partition number - but I don't see the whole of it yet and I hope there is a cleaner solution :) Let's discuss tomorrow :)

dain committed:
Add ArbitraryOutputBuffer

Commit:
Add CLI changes for password-based authentication

Add a new "password" CLI option for the user. The password
must come from a prompt and cannot be passed as an option.

Commit:
Implement LDAP authentication

On the server side:
You can have a custom user bind string property
(authentication.ldap.user-bind-pattern), which
will be used to construct the bind name to query LDAP.
This pattern *must* have a string '${USER}' as a placeholder
for the username and all instances of this string will
be replaced by the username.

Eg: "authentication.ldap.user-bind-pattern=${USER}@org.com"
with a username of 'test' will construct a bind name of "test@org.com".

Once the user's password is validated, the user's credentials will be
used to check if the user belongs to a particular group. The server
property 'authentication.ldap.group-auth-pattern' must have the group
check LDAP query.

On the CLI:
 --password flag will prompt for a password for the user

Commit:
Add OpenLDAP tests docker profile for product-tests

A new profile is created for testing LDAP authentication.
Added a basic infrastructure to support writing LDAP tests
running against OpenLDAP docker container.

sopel39 committed:
Decimal inequality operators benchmark

Commit:
Add LDAP group membership product tests

Added LDAP group membership tests along with the
OpenLDAP object definitions requirements.

sopel39 committed:
Add decimal to decimal casts benchmark

losipiuk committed:
Add micro benchmark for ORC decimal reader

pnowojski committed:
Add benchmark for decimal addition

pnowojski committed:
Add decimal benchmarks for casts

pnowojski committed:
Fix and add more tests for DOUBLE to DECIMAL casts

Commit:
Add LDAP authentication tests for Simba JDBC driver

Commit:
Add benchmarks for multiply, divide and modulo operators

pnowojski committed:
Add SingleStreamSpiller

pnowojski committed:
Add PartitioningSpiller

Partitioning spiller is required for probe pages spilling in join.
Multiple parallel operators have to spill pages to one single spill.

pnowojski committed:
Implement BinaryFileSingleStreamSpiller

pnowojski committed:
Implement GenericPartitioningSpiller

pnowojski committed:
Add getPages method to PagesIndex

maciejgrzybek committed:
Add beginQuery(Session) notification to SPI

This notification is triggered when SELECT or INSERT query is transitioning to STARTING state.

Commit:
Add password authentication tests for LDAP

pnowojski committed:
Pass spiller parameters to HashBuildOperator

pnowojski committed:
Add tests for spilling join

pnowojski committed:
Add MappedBlock

Block that maps any underlying block and automatically translates positions

pnowojski committed:
Support spilling in LookupJoinOperator

maciejgrzybek committed:
Add cleanupQuery notification to the SPI

This notification is syncrhonously called when a query is transitioning to FINISHED/FAILED state
It is called once for each catalog involved in query execution.
What is more, it's called after all other notifications (finishInsert etc.).

pnowojski committed:
Spill build table in ParallelHashBuildOperator

pnowojski committed:
Unspill build table in PartitionedLookupSource

sopel39 committed:
Add note about decimal semantics

sopel39 committed:
Precompute all possible short decimal tenToNth powers

This is required before making short decimal have 18-digit max precision

pnowojski committed:
Drop unused probeHashEnabled parameter

pnowojski committed:
Extract LookupJoiner class

cawallin committed:
Fix varbinary spilling test to use varbinary blocks

cawallin committed:
Rename spilledBytes to totalSpilledBytes

cawallin committed:
Add missing @Override to SpillerFactoryWithStats

fiedukow committed:
Add literal parameters constraints support to aggregates

fiedukow committed:
Add AggregationState annotation to avoid hardcoded 0 index

Commit:
Introduce cache for LDAP authentication

Cache authentication per credentials. It is important to cache authentication,
as we are going to use LDAP authentication for internal communication. Using LDAP
authentication without any caching will lead to severe performance degradation.

Commit:
Skip POST/PUT request content before sending the challenge

https://github.com/eclipse/jetty.project/issues/938

Commit:
Add begin/cleanup query integration test

Commit:
Refactor Kerberos authentication documentation

Refactor existing doc to have TLS specific information in a separate
page. Also updated the Kerberos config changes.

Commit:
Documentation for LDAP authentication

dain committed:
Use getCatalogHandle instead of getCatalogNames where possible

dain committed:
Add access control for SHOW and listing catalogs

dain committed:
Add access control for SHOW and listing schemas

dain committed:
Add access control for SHOW and listing tables

Commit:
Add comprehensive product tests for Cassandra

Tests including
* Dataypes
* Selecting and filtering both partitioning keys and clustering keys
* self-joins
* joins with other data sources
* negative tests

kokosing committed:
Use ImmutableTpchTablesRequirements from tempto

kokosing committed:
Remove unused unmangleNames from SqlFormatter

ArturGajowy committed:
Add TestQueryPlanDeterminism

The class tests plan determinism for most AbstractTestQueries

ArturGajowy committed:
Add IdentityLinkedHashMap

ArturGajowy committed:
Test that plan for TPC-H Q6 equivalent is deterministic

ArturGajowy committed:
Test that plan for TPC-H Q9 is deterministic

ArturGajowy committed:
Use deterministic identity Maps and Sets in planner and analyzer

Deterministic in regard to iteration order, by replacing
IdentityHashMaps by IdentityLinkedHashMap-s.

Deterministic sets are obtained by calling newSetFromMap()
on a deterministic Map.

This ensures plan determinism for the TPC-DS Q6 test case
and 4 queries from AbstractTestQueries.

ArturGajowy committed:
Fix outdated comment in EqualityInference

ArturGajowy committed:
Add PlanDeterminismChecker

pnowojski committed:
Enable join spilling in LocalExecutionPlanner

Commit:
Add zip_with function

Commit:
Introduce speculative execution properties for Cassandra

Commit:
Cleanup after tests finish in TestGrantRevoke

Commit:
Use lambda expressions in CassandraSession

Commit:
Remove unused methods from CassandraSession

Commit:
Add product tests to verify owner's privileges

kokosing committed:
Update to Airbase 60

kokosing committed:
Use directExecutorService instead of sameThreadExecutor

sameThreadExecutor is deprecated in Guava 20.

kokosing committed:
Use lambda in Threads utility

ArturGajowy committed:
Document limitations of InPredicate-dependent queries

fiedukow committed:
Avoid unannotated state arguments in aggregates

State as first argument is still supported but it's
considered legacy, so Presto functions are updated
to the new standard.

arhimondr committed:
Remove Cassandra Thrift dependency

Commit:
Remove describe_splits_ex dependency

This method from Cassandra thrift API, which is deprecated by itself, is deprecated.

Starting from Cassandra 2.1.x the "system.size_estimates" table should be used instead
to retrieve token range size estimates. Based on this estimates it is possible to split
token range into smaller sub ranges to satisfy the max_split_size criteria.

Commit:
Remove Thrift related configuration properties

dain committed:
Make PageReference a top level class

dain committed:
Simplify PartitionedOutputBuffer buffered size calculation

Use the memory manager buffer size instead of summing buffer data.

dain committed:
Simplify flushing check in output buffers

dain committed:
Rename FIXED_RANDOM_DISTRIBUTION to FIXED_ARBITRARY_DISTRIBUTION

miniway committed:
Free driver memory at Driver destroy only

losipiuk committed:
Fix building TupleDomain for comparisons with coercions

This patch fixes building TupleDomain for comparison expressions
where symbol-side of expression is wrapped with Cast

```
E.g. CAST(s AS BIGINT) >= 100000000000;
```

where s is INTEGER column

Algorithm is as follows:

We assume same expressions of same type on both sides of Comparison
expression.
First we do expression normalization which determines if
one side of comparison expression is foldable to literal and folding is
performed.
And that other side of comparison expression is either literal or
sequence of casts:
 CAST(CAST .... (CAST(symbol))))

During normalization we validate that all CASTS are coercions and bail
out if that is not the case.

Then.

If expression is of form:
```
  symbol OP literal
```
we build TupeleDomain of that.

If expression is of form:
```
  CAST(exp as T) OP literal
```

We rewrite expression to other shape which peels off one CAST and run
processing on rewritten expression. Saturated_floor_cast(litear, T) is
used in process of rewriting.

findepi committed:
Raise required Java version to 8u92

findepi committed:
Advise to use -XX:+ExitOnOutOfMemoryError in docs

The flag is available since JDK 8u92. This is 9 months old already, so
it's safe to use.

Commit:
Augment StateCompiler to support multiple variable-sized states

* Use RowType as SerializedType for states with more than two fields
* Support Block type in serializer
* Remove StateCompilerUtils

Commit:
Add Aggregation function state interfaces with two state fields

Add interfaces for two-state states:
  First state type: boolean, long, double, Slice, void/unknown
  Second state type: boolean, long, double, Slice, Block, void/unknown

Commit:
Add two-state states lookup map

Commit:
Reduce memory usage for min_by and max_by

* Generate state factory and serializer use StateCompiler
* Geneate bytecode for input, combine and output according to input type and min/max

Commit:
Add CassandraTokenSplitManager integration test

dain committed:
Move TABLE SAMPLE test to TPCH distributed queries

Table sample is handled by Presto internally by sampling splits, and for this
test to work the connector must produce multiple splits.  The TPCH connector
always produces multiple splits, so testing with just this connector is enough
to test the TABLE SAMPLE system.

dain committed:
Add optimizer to replace a 0% sample with an empty values

dain committed:
Add optimizer to remove 100% sample node

sopel39 committed:
Introduce OrderBy AST node

OrderBy node enables specifying a separate
scope for sort expressions

losipiuk committed:
Require matching types for ComparisonExpression in DomainTranslator

Add precondition check in DomainTranslator that both sides of
processed ComparisonExpression have same type. As during query
optimization it is used with expressions in such shape (any needed
coercions are already in plase as Cast expressions).

Commit:
Upgrade docker-images version to 14

Updated the kerberos docker image to generate principal with the fqdn.
Hence update the existing kerberos profile.

Commit:
Use relative temporary directory for ViewFileSystem

sopel39 committed:
Remove GetScopeVisitor

GetScopeVisitor was't used actually used in code.
Additionally, it's doesn't correctly return
scope to resolve against for expression.

sopel39 committed:
Remove unused approximate field from Scope.Builder

sopel39 committed:
Remove unneeded analysis.setScope for SubqueryExpression

sopel39 committed:
Rename and make createScope methods private

electrum committed:
Remove checkType utility methods

Due to improved generics inference in Java 8, the method is not
type safe at all, as it allows passing completely unrelated objects.
A normal cast is safer and provides a reasonable error message.

dain committed:
Fix benchmarks

kokosing committed:
Move tempto cassandra configuration

Move cassandra configuration to default tempto-configuration.yaml in
resources file. Previously it was only configured for running tempto in
docker, now it is possible to run tempto natively (without using
additional configuration from conf/tempto).

kokosing committed:
Add TPC-DS queries as product tests

The tests take about two hours to run, so disable for Travis.

kokosing committed:
Enable reorder-joins for product tests

kokosing committed:
Quarantine failing TPC-DS queries

Quarantine TPC-DS queries which do not pass on travis on master branch.
 - ones which should return CHAR type
 - ones which are using not yet merged support for GROUPING

Additionally mark with big_query the query which run over 30 minutes.

kokosing committed:
Quarantine tests that fail due to 'null IN subquery'

sopel39 committed:
Remove unnecessary setScope on ExistsPredicate

sopel39 committed:
Move StatementAnalyzer logic to inner class

Moving logic to inner class allows to have
clean StatementAnalyzer entry points.

Commit:
Extract KerberosConfig from SecurityConfig

Introduced a new property called "http-server.authentication.type"
which can have values NONE, KERBEROS (and LDAP- separate commit).

Commit:
Fix cumulative memory counter to track query memory

Previously, this counter tracked system memory, which is incorrect.

martint committed:
Update to airlift 0.140

Improves performance and memory usage patterns for QuantileDigest. It requires
adjustments due to changes to QuantileDigest serialization API.

Commit:
Update Cassandra connector documentation

electrum committed:
Cleanup code in TestingProcedures

electrum committed:
Use type names for procedure argument definitions

This avoids needing TypeManager when declaring procedures with
complex types such as array or map.

losipiuk committed:
Add TupleDomainTranslator test with multiple implicit casts

losipiuk committed:
Add TupleDomainTranslator tests with non-coercion casts

Commit:
Add copy button for query stack trace in Web UI

Commit:
Cap stack trace width on query details page

Commit:
Move resource group to execution section of query page

Commit:
Remove redundant stream instantiation for forEach

Commit:
Migrate TransformExistsApplyToScalarApply to iterative optimizer

Commit:
Copy live plan to new UI

- Migrate to using React-based updating, which makes this somewhat
  consistent with the rest of the UI
- Update styles to match new UI

We will remove the old live plan UI once this is mature.

Commit:
Clean up presto.css and utils.js

pnowojski committed:
Increase memory limit before spill

electrum committed:
Enable ALTER TABLE tests for Hive connector

Commit:
Add remote client address to query details page

Commit:
Delete old Web UI

Now that the new Web UI is mature and stable, we no longer need to keep this around.

chappidim committed:
Support KMS-managed keys S3 server-side encryption

martint committed:
Add AggregationNode.Aggregation

This class encapsulates the parts needed to represent each aggregation
function invocation (expression, signature and mask).

Currently, callers have to build three parallel maps, which makes their
code more convoluted and harder to read. With this change, they'll
be able to operate on a single map.

martint committed:
Migrate away from deprecated methods in SimplifyCountOverConstant

martint committed:
Migrate away from deprecated methods in SingleMarkDistinctToGroupBy

maciejgrzybek committed:
Add missing test for depending window functions

In case when one of two adjacent window functions
depend on the output of another WindowNodes should not be merged.

This test covers issue #7247

Commit:
Add a keystore file with a wildcard character for multinode SSL/TLS

cberner committed:
Add more documentation to InternalResourceGroup

maciejgrzybek committed:
Don't merge windows when one depends on another

Windows won't be collected once a dependency is found.
Instead, they are dumped in the plan and collection starts over.

This may result in suboptimal plan but at least will be correct.
E.g. of suboptimal solution:
There are given windows A, B, C, D.
A and C have a specification X, B and D - specification Y.
Window A depends on an output of window B, C depends on an output of D.
For such a setup the sequence of: A -> B -> C -> D could produce an optimal plan of AC -> BD.
While with this patch it will result in A -> B -> C -> D (nothing merged) instead.
This is because after seeing A, the dependent B is next, causing to dump "stashed" A.
After B, C comes but they have different specifications, therefore - both are stashed for future Window Nodes to be merged with.
Then D appears in plan which is a dependency for already "stashed" C which causes dumping the stash.
Then, the plan looks like mentioned above.

If the order in plan was different though, e.g.: A -> C -> B -> D then the final plan would be: AC -> BD (both merges possible).

This is greedy approach but solves an issue with merging dependent windows (#7247).

Commit:
Fix checkstyle violation

Commit:
Fix product tests debug environment

Local environment for product tests debugging doesn't start, because cassandra connector
is not registered, when it is used by cassandra catalog.

maciejgrzybek committed:
Disable reorder-windows by default to prevent a correctness error

In corner case mentioned in #7247 the reorder-windows optimizer may result in incorrect plan.
Once patch is prepared, the optimizer will be enabled back.

Commit:
Handle escape in LIKE predicate optimization

Commit:
Update for tempto 1.26

electrum committed:
Remove redundant strings from PrestoS3SignerType

findepi committed:
Remove stuttered words from comments and documentation

Commit:
Fix hive.config.resources property behaviour

`hive.config.resources` property allows to set additional resources
that must override the default properties for HDFS.

`Configuration#addResource` method loads the resource, and handles the
properties from that resource kind of like a defaults. So if before
some property has been set explicitly via `set*` method, that property
won't be overwritten with the value loaded from the resource.

The way how defaults are being populated to the configuration uses
`set` method. So, further properties loaded from the specified resources
can't override the defaults.

This commits loads the resource into temporary clean configuration. And
then copies all the properties to the destination using the `set` method,
so that way the defaults are being actually overwritten.

Commit:
Handle escape correctly in like predicate and LikeFunctions

wenleix committed:
Fix bad type signature generated by optimizer

OptimizeMixedDistinctAggregations generates bad type signature for nested row types.

martint committed:
Update to airlift 0.141

Fixes a thread-safety issue in the copy constructor of Distribution

kokosing committed:
Do not require all TPC-H tables if they are not used

findepi committed:
Fix implementation of DereferenceExpression.getChildren()

electrum committed:
Update to Airbase 61

Commit:
Add function to compute Levenshtein distance between two strings

Add new function levenshteinDistance to StringFunctions to compute the edit distance
between two strings. A new unit test is added with patch.

Commit:
Refactor to inject OrderingCompiler and JoinCompiler

Refactor various classes to inject singletons of these two classes
to be able to track the cache hit/miss metrics properly.

haozhun committed:
Fix getSizeInBytes in ArrayBlock

electrum committed:
Add debug logging for class definition

haozhun committed:
Add test for Block.getSizeInBytes

Commit:
Add pipeline ID to PipelineContext

An operator ID is not enough to uniquely identify an operator in a
stage. Introducing a pipeline ID solves this problem, under the
assumption that all workers will generate pipeline IDs in the same
order.

Commit:
Add react-dom-server dependency

React DOM Server is used to manually render DagreD3 graph nodes.

Commit:
Change rendering of live plan node to be React based

Also factor out some functionality into utils.js

Commit:
Clean up and unify Web UI query navigation

Commit:
Add stage performance page to Web UI

This page helps understand operator-level performance in every stage
and pipeline. It also provides a view of operator performance across
tasks to quickly find skew.

Commit:
Fix formatDataSize to include min unit

Commit:
Expose JoinCompiler cache stats via JMX

Commit:
Expose OrderingCompiler cache stats via JMX

Commit:
Expose JoinFilterFunctionCompiler cache stats via JMX

Commit:
Refactor HiveHdfsConfiguration

Extract Configuration properties copying into utility method

Commit:
Read HDFS configuration at startup

Parsing of the XML files might be expensive. It is better to read them
once, and then save the result in memory.

Commit:
Unify Cassandra Datastax Driver version

In presto-cassandra and presto-product-tests modules

Commit:
Do not expose Thrift port in Cassandra product tests

Commit:
Update Cassandra to 3.9 in Product Tests

Commit:
Upgrade query details page to use ES6 features

electrum committed:
Use shaded Hadoop dependency for Accumulo connector

electrum committed:
Update ZooKeeper client to 3.4.9

findepi committed:
Improve handling of missing Fields in RelationType.indexOf

Previously, the method would throw NullPointerException without a
message when passed field was null or not found, although documentation
stated differently.

Now exception is always with a message, and documentation is aligned
with the code.

findepi committed:
Include original exception when it has wrong message

Commit:
Update Tempto to 1.27

Commit:
Short-circuit inner and right join when right side is empty

electrum committed:
Improve error handling when Hive partitions already exist

If we get an AlreadyExists error from the metastore, we don't know which
partitions already exists, so indicate that in the error message. Also
include the original metastore error in case it has helpful information.

Commit:
Make blocked state easier to spot

martint committed:
Update to airlift 0.142

Fixes a contention issue in Distribution's copy constructor
that can cause coordinator->worker requests to get "stuck"
computing various stats.

Commit:
Wrap long user-provided strings on query overview page

haozhun committed:
Add 0.166 release notes

electrum committed:
[maven-release-plugin] prepare release 0.166

electrum committed:
[maven-release-plugin] prepare for next development iteration

Commit:
Document zip_with

electrum committed:
Remove unnecessary locking in assertQueryFails method

Locking is not needed since no QueryRunner state is changed.

wenleix committed:
Add transform_key lambda function

Commit:
Expose JoinProbeCompiler cache stats via JMX

findepi committed:
Remove unnecessary conversions from IntStream to LongStream

kokosing committed:
Remove unnecessary List.transform

erichwang committed:
Extract AllowAllSystemAccessControl as separate class

erichwang committed:
Add read-only system access controller

electrum committed:
Fix merge conflict in TestDecimalCasts

findepi committed:
Pass ESCAPE in JDBC metadata LIKE predicates

ArturGajowy committed:
Divide probe unspilling between all LookupJoinOperator-s

Each lookup-join operator now gets its own isntance of
PartitioningSpiller for spilling and unspilling parts of probe table.

ArturGajowy committed:
Make LookupSource unspilling asynchronous

ArturGajowy committed:
Static import MoreFutures.toCompletableFuture

electrum committed:
Remove unused test dependency cassandra-driver-mapping

maciejgrzybek committed:
Expose port 9000 for hadoop-master container

ArturGajowy committed:
Unspill each partition at most once

thanks to memoization of the returned CompletableFuture-s

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java:
```diff
@@ -82,6 +78,9 @@
     private final List<SettableFuture<LookupSource>> lookupSourceFutures = new ArrayList<>();
 
+    @GuardedBy("this")
+    private final Map<Integer, CompletableFuture<LookupSource>> unspillLookupSourceFuturesByPartitionCache = new HashMap<>();
```
I don't think we need this map here, since it always should contain at most one element. Having two fields: integer and one `Optional<CompletableFuture<LookupSource>>`  would be more implicit.

pnowojski replied:
what about closing those lookup sources? What happens in case of some abnormal interrupt/cancelling the query?

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java:
```diff
@@ -254,24 +245,32 @@ private void checkAllFuturesDone()
 
     @Override
-    public CompletableFuture<LookupSource> readSpilledLookupSource(Session session, int partition)
+    public synchronized CompletableFuture<LookupSource> readSpilledLookupSource(Session session, int partition)
```
besides synchronization on loading next partitions, you also need a code to clean up and close no longer used partitions.

pnowojski commented on presto-main/src/main/java/com/facebook/presto/spiller/SingleStreamSpiller.java:
```diff
@@ -44,6 +45,12 @@
+     * Returns a future computation of list of pages
+     * obtained by iterating the {@code getSpilledPages} Iterator.
+     */
+    CompletableFuture<List<Page>> getSpilledPagesFuture();
```
getSpilledPagesAtOnce?

getAllSpilledPages?

pnowojski reviewed (changes requested):
% missing synchronization and cleaning op it looks good

pnowojski committed:
Migrate PredicatePushdown#visitFilter to iterative optimizer

Commit:
Add JMX stats for longest running split elapsed time

All ongoing splits' starting time are recorded in a set in the task
executor. This provides the insight of how long these splits have been
running. A new interface is added to return the longest consecutive split
execution time.

dain committed:
Cleanup ArbitraryOutputBuffer and TestArbitraryOutputBuffer

electrum committed:
Update to hive-dwrf 0.8.2

electrum committed:
Use shaded versions of Cassandra server and driver

haozhun committed:
Refactor ParquetReader to not use ArrayBlock.getLength

haozhun committed:
Fix SliceArrayBlock.getSizeInBytes

* To not include object overhead of SliceArrayBlock
* To include object overhead of Slice

haozhun committed:
Add Block.getRegionSizeInBytes to remove improper use of getLength

This fixes Block.getSizeInBytes when ArrayBlock is in DictionaryBlock.

haozhun committed:
Rename Block.getLength to Block.getSliceLength

haozhun committed:
Remove getSliceLength from Blocks without getSlice

kokosing committed:
Rename rewriteQualifiedNamesToSymbolReferences

QualifiedNameReference no longer exists so the name of method seem to be
obsolete. This commit updates this name of this method to refer that it
rewrites Identifiers to SymbolReferences.

electrum committed:
Remove legacy Hive connectors from server tarball

raghavsethi committed:
Add stat to track quanta wall time

raghavsethi committed:
Change Web UI to load Roboto from Google CDN

dain committed:
Implement LazyOutputBuffer.getUtilization

electrum committed:
Remove support for legacy S3 Block FileSystem

findepi committed:
Convert asserts into runtime checks

findepi committed:
Forbid using assert

wenleix committed:
Add transform_value lambda function

haozhun committed:
Add comments to LookupJoinOperator

electrum committed:
Remove checkNotSameThreadExecutor

haozhun committed:
Add evaluation limit to LookupJoinOperator to yield driver thread

electrum committed:
Add synchronized PagesSerde for testing

This fixes race conditions in TestArbitraryOutputBuffer.

ArturGajowy committed:
Coordinate LookupJoinOperators so that all consume the same lookup partition

findepi committed:
Use quotes around variables in presto-launcher-wrapper.sh

Commit:
Remove running split info before completing its finishedFuture

ListenableFuture::get is listening on the finish of split. If a running
split info is removed after it has been set to finished, There may still
have remaining running split info alive. This can fail the unit test due
to race.

raghavsethi committed:
Add isQueryComplete helper method

raghavsethi committed:
Improve visibility of scheduled and blocked time

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java:
```diff
@@ -113,18 +117,9 @@ public void finish()
-                ensureSpillerLoaded();
-                unspillNextLookupSource();
-            }
+            lookupSourceFactory.beginLookupSourceUnspilling(lookupJoinsCount.get(), operatorContext.getSession());
```
If you are not decrementing, you can use `final int` instead of atomic.

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java:
```diff
@@ -135,15 +130,10 @@ private void unspillNextLookupSource()
         checkState(spilledPartitions.isPresent());
         if (spilledPartitions.get().hasNext()) {
             int currentSpilledPartition = spilledPartitions.get().next();
```
previously there was no need for a contract, that `spilledPartitions` iterator returns always elements in the same order, for each LJO. It should work fine now, but It might be better to refactor this somehow? For example obtain `currentSpilledPartition` value from `lookupSourceFactory` and let it take care of the order of unspilling?

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java:
```diff
@@ -135,15 +130,10 @@ private void unspillNextLookupSource()
-                    joinProbeFactory,
-                    spiller.get().getSpilledPages(currentSpilledPartition),
-                    probeOnOuterSide));
+            currentLookupPartition = Optional.of(lookupSourceFactory.unspillLookupPartition(currentSpilledPartition));
```
Maybe add some `checkState` before overwriting that previous partition has been fully processed? Maybe that will catch some bug?

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java:
```diff
@@ -167,6 +157,9 @@ public boolean isFinished()
             return spilledLookupJoiner.get().isBlocked();
         }
+        if (currentLookupPartition.isPresent()) {
+            return toListenableFuture(currentLookupPartition.get());
```
why do you need to block on both `spilledLookupJoiner` and `currentLookupPartition`?

pnowojski replied:
We also have to make sure that when we are logically waiting on two futures to complete, that we return combined future of both of them. If not, we could return almost finished future, that completes before caller checks `isDone` on it.

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java:
```diff
@@ -233,21 +226,41 @@ private Page mapPage(IntArrayList unspilledPositions, Page page)
     {
-        if (finishing && spiller.isPresent()) {
-            checkState(spilledLookupJoiner.isPresent());
+        if (finishing && spiller.isPresent() && currentLookupPartition.map(CompletableFuture::isDone).orElse(false)) {
```
Maybe put `checkState(!currentLookupPartition.isPresent() || currentLookupPartition.get().isDone())` before this if?

pnowojski reviewed (commented):
Sorry only partial review. I will finish it on Monday.

raghavsethi committed:
Fix checks for updating quanta stats

Corrected the blocked/unblocked check to remove the flipped
conditional and use Operator.NOT_BLOCKED.

Removed the isFinished check, as it does not appear to have any
material effect on the reported results, and has unintended side
effects (including test failures).

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java:
```diff
@@ -220,30 +220,21 @@ public void destroy()
-    }
-
-    @Override
     public synchronized PartitioningSpiller getProbeSpiller(List<Type> probeTypes, HashGenerator probeHashGenerator)
```
`createProbeSpiller`?

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java:
```diff
@@ -254,24 +245,26 @@ private void checkAllFuturesDone()
+    public synchronized void beginLookupSourceUnspilling(int consumers, Session session)
+    {
+        checkState(hasSpilled());
+        lookupSourceUnspilling.ifPresent(consumption -> checkState(consumption.getConsumersCount() == consumers));
```
since you already have if statement for this condition, it would be cleaner to have simple 
```
if (lookupSourceUnspilling.isPresent) {
  ...
} 
else {
  ...
}
```
then to mix imperative and functional programming in one logical access.

Plus, please add meaningful `checkState` fail massage (FB guys were complaining that I wasn't doing that ;) )

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java:
```diff
@@ -359,4 +352,162 @@ public SetLookupSourceResult()
         }
     }
+
+    private static class PartitionConsumption
```
naming for those classes is confusing

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java:
```diff
@@ -359,4 +352,162 @@ public SetLookupSourceResult()
+                return markRequested(currentPartition).loaded;
+            }
+            else {
+                checkArgument(partitions.hasNext(), format("Requested next partition with number [%s] while the only one left is of different number", partitionNumber));
```
error message doesn't fit condition?

ArturGajowy replied:
it does fit - see the if condition for that `else`

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java:
```diff
@@ -359,4 +352,162 @@ public SetLookupSourceResult()
+
+        private PartitionConsumption createConsumption(Integer partitionNumber)
+        {
+            CompletableFuture<Void> partitionRequested = new CompletableFuture<>();
```
I would suggest cutting down on number of futures and dependencies between them, otherwise this code will be hard to understand and to follow (both when reading and while debugging).

For example , some (most?) of those "tasks" fired on future completions could be fired explicitly by caller methods. Like calling `release` could  could check whether that was the "last user of previous partition" and if so, call `LookupSourceFactory::releasePartition(...)` which would directly trigger unspilling next partition.

ArturGajowy commented:
@pnowojski looks like I managed to fix the tests by removing the Optional-frenzy that I added in the previous version of the 'coordinate operators' commit :) It seems that we can move on to polishing :)

ArturGajowy commented:
Dropped the `Make OperatorAssertions consider Operator#isBlocked` commit as it has no effects on test results, now that we can verify it

ArturGajowy committed:
Check size is known in LookupJoinOperator#getInMemorySizeInBytes()

ArturGajowy committed:
Do lookup memory reservation per group of LookupJoinOperators

ArturGajowy commented:
Added 2 more commits, introduced OperatorSharedMemory for accounting the shared memory. Later on we might migrate it to some better implementation, but its API seems ~stable~ OK for now.

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java:
```diff
@@ -77,11 +78,12 @@ public LookupJoinOperator(
             Runnable onClose,
             AtomicInteger lookupJoinsCount,
-            HashGenerator hashGenerator)
+            OperatorSharedMemory operatorSharedMemory, HashGenerator hashGenerator)
```
formatting

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorSharedMemory.java:
```diff
@@ -0,0 +1,73 @@
+    private Long inMemorySizeInBytes;
+    private TaskContext taskContext;
+
+    public synchronized void setReservation(int partitionKey, OperatorContext operatorContext, long inMemorySizeInBytes)
```
`setBytes`

ArturGajowy replied:
setReservedBytes?

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorSharedMemory.java:
```diff
@@ -0,0 +1,73 @@
+    private Long inMemorySizeInBytes;
+    private TaskContext taskContext;
+
+    public synchronized void setReservation(int partitionKey, OperatorContext operatorContext, long inMemorySizeInBytes)
```
by looking at this class's interface I'm confused what `partitionKey` is. Maybe try to refactor it so it's more obvious? Rename it? As a last resort add java doc?

It doesn't seem to fit into more general concept of "shared memory". If you need it for assertions only, maybe achieve this in some other way? This seems like leakage of implementation detail of spilling.

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorSharedMemory.java:
```diff
@@ -0,0 +1,73 @@
+            this.partitionKey = partitionKey;
+            this.inMemorySizeInBytes = inMemorySizeInBytes;
+            this.taskContext = operatorContext.getDriverContext().getPipelineContext().getTaskContext();
+            operatorContext.setMemoryReservation(inMemorySizeInBytes);
```
this doesn't set reservation to `inMemorySizeInBytes` but always increase memory reservation by `inMemorySizeInBytes`. This is not `set***` method but `update***` method. Generally speaking `set*` are easier to use.

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorSharedMemory.java:
```diff
@@ -0,0 +1,73 @@
+                "Inconsistent memory de-reservation. " + getInconsistenceDescription(partitionKey, inMemorySizeInBytes));
+        reservations--;
+        if (reservations == 0) {
+            this.partitionKey = null;
```
do not use nulls

pnowojski reviewed (changes requested):
I would like to rereview whole change later, could you apply changes for previous comments and restructure/reorder them? 

I am not sure whether `OperatorSharedMemory` is a right direction. It adds lots of code for a stop gap solution.
